### ISSUE TITLE
Mutability of view nodes

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/sun#290]
The biggest pain point during development of the last two sims has by far been common view components not supporting mutable properties or resizing content.

Examples include:

- AccordionBox not being able to resize based on content
- RadioButtonGroup not being able to re-layout when item nodes change size
- RadioButtonGroup's colors can't be changed
- RectangularPushButton/BooleanRoundToggleButton/ResetButton's colors can't be changed
- NumberPicker's colors can't be changed
- NumberSpinner's range couldn't be changed (just refactored to a property today)
- ShadedSphereNode's colors can't be changed
- Passing in enabledProperty not supported by ButtonModel, so when buttons get recreated (due to a color change) extra code is required.
- Keypad layout (or button visibility) changes not possible

I've been working around these by having to rebuild these components (usually with MutableOptionsNode) or contortions to compute the ideal size of the accordion/radiobuttongroup content before creation.

This breaks down when I want to, for example, change a NumberPicker's color during a pointer press, as recreating would presumably interrupt the press (anticipated need in Area Model).

I'd like to propose:

1. New common components should have this type of mutation as the "default" (with a {Property}, getter/setter, whatever) for options, unless it doesn't make sense (i.e. "no one would want this, I'll buy JO a six-pack of beer if he needs to mutate this").
2. The above examples get fixed to include mutability (and the option to resize, by default off for compatibility). I'm willing to do so if @ariel-phet says I have the time.

**Comment by samreid** [comment]
I haven't had any sims where those described mutations would be necessary but it's easy to imagine cases where they would be valuable and I'm not opposed to having them added.  Also, let's consider `tandem` names and make sure this kind of mutation is not overdone.  For instance, the tandem/phet-io interface (and the sim implementation) would be awkward if there is a red "add electron" button that gets mutated into a blue "delete skateboarder" button--in that case it would be clearer to have two different buttons.

Also, when you say resize, are you referring to internal padding within components, scaling within components or something else?

**Comment by jonathanolson** [comment]
> Also, when you say resize, are you referring to internal padding within components, scaling within components or something else?

An actual Area Model example that inspired this issue:

I have two RadioButtonGroups that are in two different panels (in two different files). They have icons that are specific to each file. It's now desired to have those buttons be the same size across the two panels.

- Desired implementation: I pass in an AlignGroup to both panels for the icon sizing. Each panel creates its icons, wraps them in an AlignBox referencing the group, and no more effort is needed (all icons thus have the necessary amount of padding so that the buttons are the same size).
- Why this doesn't work: because the icons in the first panel are not as large as the largest icon, the RadioButtonGroup in the first panel gets constructed with a smaller button size. When the icon sizes are increased (when the second panel wraps its icons in the AlignBoxes), the first RadioButtonGroup doesn't re-layout or re-size based on the new content sizes.
- What I'd have to do to work around: I can't create the icons where they are used any more, since ALL of the icons need to be created (and AlignBoxed or otherwise have maximum dimensions computed) before ANY RadioButtonGroup is created. Instead, I'll need to have a third file which knows how to create all of the icons and does so. BUT it gets even worse, since the icons' appearance isn't static, and needs to be hooked to model properties. Things end up with details/state getting leaked everywhere, having to create "fake" model properties to pass in just to get sizing, etc. A 1-minute refactor (pass an align group through to each panel) turns into a monster.

**Comment by jonathanolson** [comment]
> Also, let's consider tandem names and make sure this kind of mutation is not overdone.

Is it awkward for phet-io if view nodes for the same model object are getting constantly replaced with slightly different view nodes every time a model property changes? That would presumably cause a lot of tandem.addInstance / tandem.removeInstance for something that shouldn't even be changing.

**Comment by samreid** [comment]
> Is it awkward for phet-io if view nodes for the same model object are getting constantly replaced with slightly different view nodes every time a model property changes? 

Yes, PhET-iO interface is simpler for static items.  By static, I mean that `tandem.addInstance` is only called during simulation startup, and `tandem.removeInstance` is never called.

**Comment by pixelzoom** [comment]
I understand the frustrations and motivation for suggesting mutability of UI components.  

A few additional thoughts and observations:

(1) It's often unclear when implementing (or generalizing) a UI component which aspects need to be mutable. So when @jonathanolson says "New common components should have this type of mutation as the default", that sounds nice. But what exactly is "this type of mutation"?

(2) Each aspect of a UI component that is mutable adds incremental implementation expense. Making everything mutable is cost-prohibitive, and likely wasteful (some mutability may never be used).  How to we identify what should and should not be mutable? Should certain aspects be mutable in all PhET UI components?

(3) Mutable implementations have proven to be more complicated, and therefore more expensive to implement, extend and maintain.  For example, HSlider is already huge - I'd hate to see what it looks like after making lots of things mutable.

(4) Creating a new view (rather than mutating an existing view) _sometimes_ results in (a) simplified client-side implementation and (b) better performance.  

(5) It's implied above that persistent (mutable) UI components are preferred for PhET-iO.  If that's the case, then that needs to be considered when making software architecture decisions.   And it will certainly have (expensive!) implications for adding instrumentation to existing sims that don't follow the "persistent UI" pattern.

**Comment by jonathanolson** [comment]
> It's often unclear when implementing (or generalizing) a UI component which aspects need to be mutable.

> Each aspect of a UI component that is mutable adds incremental implementation expense. 

Agreed that we don't want to over-do mutability.

> Mutable implementations have proven to be more complicated, and therefore more expensive to implement, extend and maintain.

Understood, but using workarounds (and then having to rewrite into a mutable form) also classifies the same way.

> Creating a new view (rather than mutating an existing view) sometimes results in (a) simplified client-side implementation and (b) better performance.

Agreed, it can be a balancing act.
